omniledger/service/service.go

```diff
@@ -301,31 +299,43 @@ func (s *Service) createNewBlock(scID skipchain.SkipBlockID, r *onet.Roster, cts
 	// Create header of skipblock containing only hashes
 	var scs StateChanges
 	var err error
-	var ctsOK ClientTransactions
+	var ctsOK, ctsBad ClientTransactions
 
 	log.Lvl3("Creating state changes")
-	mr, ctsOK, _, scs, err = s.createStateChanges(coll, scID, cts, noTimeout)
+	mr, ctsOK, ctsBad, scs, err = s.createStateChanges(coll, scID, cts, noTimeout)
 
 	if err != nil {
 		return nil, err
 	}
 	if len(scs) == 0 {
 		return nil, errors.New("no state changes")
 	}
+
+	refused := make([]ClientTransactionID, len(ctsBad))
+	rhash := sha256.New()
+	for i, bad := range ctsBad {
+		refused[i] = bad.Instructions.Hash()
+		rhash.Write(refused[i])
+	}
+
 	header := &DataHeader{
-		CollectionRoot:        mr,
-		ClientTransactionHash: ctsOK.Hash(),
-		StateChangesHash:      scs.Hash(),
-		Timestamp:             time.Now().UnixNano(),
+		CollectionRoot:               mr,
+		ClientTransactionHash:        ctsOK.Hash(),
+		ClientTransactionRefusedHash: rhash.Sum(nil),
+		StateChangesHash:             scs.Hash(),
+		Timestamp:                    time.Now().UnixNano(),
 	}
-	sb.Data, err = network.Marshal(header)
+	sb.Data, err = protobuf.Encode(header)
 	if err != nil {
 		return nil, errors.New("Couldn't marshal data: " + err.Error())
 	}
 
 	// Store transactions in the body
-	body := &DataBody{Transactions: ctsOK}
-	sb.Payload, err = network.Marshal(body)
+	body := &DataBody{
+		Transactions:        ctsOK,
+		RefusedTransactions: refused,
+	}
```

Comment:
This part doesn't look right. Don't we want to store all the transactions but mark the ones that are bad? The follower should all execute every transaction and find the same transactions to be bad. I don't think the actual bad transactions are sent out for the followers to check.

Reply:
During the design discussion for this we determined that allowing an attacker to fill the body with bad tx's was not a good plan, so we'd only store the hashes.

Reply:
On one of the first implementations we had a `bool` field for every `ClientTransaction` in the body of the block. That would make sense here.

Reply:
On further consideration we found that it is imperative to have the tx+bool setup, because that way the block verifiers get a chance to run the exact same txns in the same order as the leader and see if they also detect the same set of accepted/refused txns.